src/frontend/src/expr/mod.rs

```diff
@@ -174,21 +183,34 @@ impl ExprImpl {
     /// Check whether self is a literal NULL or literal string.
     pub fn is_unknown(&self) -> bool {
         matches!(self, ExprImpl::Literal(literal) if literal.return_type() == DataType::Varchar)
+            || matches!(self, ExprImpl::Parameter(parameter) if !parameter.has_infer())
     }
 
     /// Shorthand to create cast expr to `target` type in implicit context.
     pub fn cast_implicit(self, target: DataType) -> Result<ExprImpl, CastError> {
-        FunctionCall::new_cast(self, target, CastContext::Implicit)
+        if let ExprImpl::Parameter(expr) = &self && !expr.has_infer() {
+            Ok(expr.to_owned().cast_infer_type(target).into())
```

Comment:
Is it possible to put these repeated logic inside `FunctionCall::new_cast`?

Reply:
I think yes. But whether it conform to the semantics of FunctionCall::new_cast🤔, it did't create a functioncall actually.

Reply:
Right. `FunctionCall::new_cast` does not always creates a `FunctionCall`. There are already some simplification inside `FunctionCall::new_cast`, like skip casting when type already matches, or handling `null` and parsing. It may be better to put this together with them, and maybe refactor all of them together later.